edgelet/edgelet-docker/src/image_prune_data.rs

```diff
@@ -0,0 +1,635 @@
+                return Ok(HashMap::new());
+            }
+        };
+
+        /* ============================== */
+
+        // process maps
+        let (images_to_delete, carry_over) = process_state(
+            iotedge_images_map,
+            in_use_image_ids,
+            settings.image_age_cleanup_threshold(),
+        )?;
+
+        /* ============================== */
+
+        // write previously removed entries back to file
+        if let Err(e) = write_images_with_timestamp(
+            &carry_over,
+            guard.tmp_filepath.clone(),
+            guard.image_use_filepath.clone(),
+        ) {
+            log::warn!("Failed to update image auto pruning persistence file. File will be updated on next scheduled run. {}", e);
+        };
+
+        /* ============================== */
+
+        drop(guard);
+
+        // these are the images we need to prune; file has already been updated
+        Ok(images_to_delete)
+    }
+}
+
+/* ===================================== HELPER METHODS ==================================== */
+
+fn get_images_with_timestamp(
+    image_use_filepath: String,
+) -> Result<HashMap<String, Duration>, Error> {
+    if !std::path::Path::new(&image_use_filepath).exists() {
+        log::info!(
+            "Image garbage collection data file not found; creating file at: {}",
+            image_use_filepath.as_str()
+        );
+        let _file = fs::File::create(image_use_filepath.clone()).map_err(Error::CreateFile)?;
+    }
+
+    let res = fs::read_to_string(image_use_filepath);
+    if let Err(e) = res {
+        let msg = format!("Could not read image persistence data: {}", e);
+        log::error!("{msg}");
+        return Err(Error::FileOperation(msg));
+    }
+
+    let contents = res.expect("Reading image persistence data failed");
+
+    let mut image_map: HashMap<String, Duration> = HashMap::new(); // all image pruning data
+
+    // TL;DR: this dumps pruning data into the image_map, where
+    // Key: Image hash, Value: Timestamp when image was last used (in epoch)
+    let contents = contents
+        .lines()
+        .map(|line| line.split(' ').collect::<Vec<&str>>())
+        .map(|vec| (vec[0].to_string(), vec[1]));
+    for (k, v) in contents {
+        image_map.insert(
+            k,
+            Duration::from_secs(v.parse::<u64>().map_err(Error::ParseIntError)?),
+        );
+    }
+
+    Ok(image_map)
+}
+
+fn write_images_with_timestamp(
+    state_to_persist: &HashMap<String, Duration>,
+    temp_file: String,
+    image_use_filepath: String,
+) -> Result<(), Error> {
+    // write to a temp file and then rename/overwrite to image persistence file (to prevent file write failures or corruption)
+    let mut file = std::fs::File::create(temp_file.clone())
+        .expect("Could not create temporary persistence file");
+
+    for (key, value) in state_to_persist {
+        let image_details = format!("{} {}\n", key, value.as_secs());
```

Comment:
Use serde to serialize/deserialize instead of implementing it manually. Choose a format (json/toml/etc) and use the corresponding serde library to convert.

Reply:
@gordonwang0 - You're right. This is a great suggestion, and I should have done it this way.

That said, this suggestion is coming a bit too late and this is a much-awaited feature for the LTS release. So even though I agree with you, I'm not inclined to take this suggestion because this causes a lot of churn at this very late stage.

Mitigating factors include:

- this is an implementation detail (that's easily changed in the future), and not something the users need to worry about.
- this feature is somewhat orthogonal to what iotedge does and this code is unlikely to be touched often enough to be any kind of a maintenance burden (and if it is, we can calibrate accordingly and make the changes you're proposing).


Any time spent on this feature at this juncture should be on testing, and not on rewriting something that works, 

Again - You're right. Completely. But we don't have enough time if we want to test this and get LH kicked off soon. (unless someone who is has deeper knowledge of serialization libraries - and rust in general - than I, is willing to make this change and make a push).

Reply:
Ok, that's fine.